design/ceph/multus-network.md

```diff
@@ -78,7 +78,123 @@ Nothing to do in particular since they don't use any service IPs.
+application. We can document for users that they should not delete holder pods, but we cannot
+prevent users from manually stopping holder pods. However, to prevent the Kubernetes scheduler from
+terminating holder pods, the pods should be given the highest possible priority so that they are not
+un-scheduled except by users.
+
+There is a possible race condition where a mover pod is killed and where a holder is deleted before
+a new mover starts up. In this case, the copied interface for the holder pod will be left in the
+host network namespace, but the mover will not get a notification that the holder was removed. In
+order to clean up from this case, upon startup, the mover should delete all interface copies in the
+host network namespace that do not have a holder pod associated with them.
+
+If the mover container is stopped, it should delete all copied interfaces in the host's
+network namespace under the assumption that the CSI plugin is being removed, possibly by a
+CephCluster being deleted or the node going down for maintenance. It might be possible to optimize
+for the case where only the mover pod is being restarted, but it is very difficult to detect the
+case where the mover is merely being restarted versus when the holder is also being removed without
+possible race conditions between which pod might be stopped first by the Kubernetes API server
+during a drain event. Therefore, focus on the simplest working implementation (described above)
+instead of risking leaving the interface copy in host net namespace which could cause issues.
+
+If an error occurs in the mover during network migration, it will fail and re-try migration until
+the operation succeeds. If necessary, the mover will try to remove a partially-copied interface.
+
+Restarting a node will cause the multus interface in the host namespace to go away. On restart, the
+holder pod will get a new interface, and the mover will copy it into the host networking namespace again.
+
+When a new node is added, holder and mover pods are added to it by their DaemonSets, and the
+move/copy process described above occurs on the node.
+
+The holder and mover DaemonSets should be deleted when the CSI driver components are removed. Both
+the termination of the holder pods as well as the mover pods triggers the mover to remove the multus
+interfaces from the host network namespace of a given node.
+
+The initial implementation of this design will be limited to supporting a single CephCluster with
+Multus until we can be sure that the CSI plugin can support multiple migrated interfaces as well as
+interfaces that are added and removed dynamically. This limitation will be enforced by allowing only
+a single instance of the holder DaemonSet. A possible (future) partial implementation may be
+possible by restarting the CSI plugin Pods when network interfaces are added or removed.
+
+**Known issue:** the Docker container runtime does not use Linux's native `/var/run/netns`
```

Comment:
We could make this path configurable too.

Reply:
Ideally we would, and I hope we can in the future to support Docker. We would have to mount a different dir into the mover container, and we might also have to update SecurityContextConstraints and/or PodSecurityPolicies (or PSPNext) to allow access to different directories, which complicates things for users a bit.